tests/EventTests/EventBuilderTest.php

```diff
@@ -309,6 +309,100 @@ public function testCreateImpressionEventWithUserAgentWhenBotFilteringIsEnabled(
         $this->assertTrue($result[0], $result[1]);
     }
 
+    public function testCreateImpressionEventWithAttributesOfValidType()
+    {
+        array_unshift($this->expectedEventParams['visitors'][0]['attributes'],
+            [
+                'entity_id' => '7723280020',
+                'key' => 'device_type',
+                'type' => 'custom',
+                'value' => 'iPhone',
+            ],[
+                'entity_id' => '7723340007',
+                'key' => 'boolean_key',
+                'type' => 'custom',
+                'value' => true
+            ],[
+                'entity_id' => '7723340008',
+                'key' => 'double_key',
+                'type' => 'custom',
+                'value' => 5.5
+            ],[
+                'entity_id' => '7723340009',
+                'key' => 'integer_key',
+                'type' => 'custom',
+                'value' => 5
+            ]
+        );
+
+        $this->expectedLogEvent = new LogEvent(
+            $this->expectedEventUrl,
+            $this->expectedEventParams,
+            $this->expectedEventHttpVerb,
+            $this->expectedEventHeaders
+        );
+
+        $userAttributes = [
+            'device_type' => 'iPhone',
+            'boolean_key' => true,
+            'double_key' => 5.5,
+            'integer_key' => 5
+        ];
+        $logEvent = $this->eventBuilder->createImpressionEvent(
+            $this->config,
+            'test_experiment',
+            'variation',
+            $this->testUserId,
+            $userAttributes
+        );
+
+        $logEvent = $this->fakeParamsToReconcile($logEvent);
+        $result = $this->areLogEventsEqual($this->expectedLogEvent, $logEvent);
+        $this->assertTrue($result[0], $result[1]);
+    }
+
+    public function testCreateImpressionEventWithAttributesOfInValidType()
```

Comment:
nit. WithInvalidAttributeTypes

Reply:
Also add a comment that verify that only valid attributes are part of the generated payload.